pgm/rawtopgm.1
.TH rawtopgm 1 "14 September 2000"
.IX rawtopgm
.SH NAME
rawtopgm - convert raw grayscale bytes into a portable graymap
.SH SYNOPSIS
.B rawtopgm
.RB [ "-bpp " [ 1 | 2 ]]
.RB [ -littleendian ]
.RB [ -maxval
.IR N ]
.RB [ -headerskip
.IR N ]
.RB [ -rowskip
.IR N ]
.RB [ -tb | -topbottom ]
.RI [ width
.IR height ]
.RI [ imagefile ]

.SH DESCRIPTION
Reads raw grayscale values as input.
.IX "raw grayscale"
Produces a PGM file as output.  The input file is just a sequence of
pure binary numbers, either one or two bytes each, either bigendian or
littleendian, representing gray values.  They may be arranged either
top to bottom, left to right or bottom to top, left to right.  There
may be arbitrary header information at the start of the file (to which
.B rawtopgm
pays no attention at all other than the header's size).

Arguments to 
.B rawtopgm
tell how to interpret the pixels (a function that is served by a header in
a regular graphics format).

The 
.I width
and
.I height
parameters tell the dimensions of the image.  If you omit these parameters,
.B rawtopgm
assumes it is a quadratic image and bases the dimensions on the size of the
input stream.  If this size is not a perfect square, 
.B rawtopgm
fails.

When you don't specify
.I width
and
.IR height ,
.B rawtopgm
reads the entire input stream into storage at once, which may take a
lot of storage.  Otherwise,
.B rawtopgm
ordinarily stores only one row at a time.

If you don't specify 
.IR imagefile ,
or specify 
.BR - ,
the input is from Standard Input.

The PGM output is to Standard Output.



.SH OPTIONS
.TP
.B -maxval \fIN
.I N
is the maxval for the gray values in the input, and is also the maxval
of the PGM output image.  The default is the maximum value that can be
represented in the number of bytes used for each sample (i.e. 255 or
65535).
.TP
.B -bpp \fR[\fB1\fR|\fB2\fR]
tells the number of bytes that represent each sample in the input. 
If the value is 
.BR 2 ,
The most significant byte is first in the stream.

The default is 1 byte per sample.

.TP
.B -littleendian
says that the bytes of each input sample are ordered with the least
significant byte first.  Without this option, 
.B rawtopgm
assumes MSB first.  This obviously has no effect when there is only one byte
per sample.

.TP
.B -headerskip \fIN
.B rawtopgm
skips over 
.I N
bytes at the beginning of the stream and reads the image immediately after.
The default is 0.

This is useful when the input is actually some graphics format that has
a descriptive header followed by an ordinary raster, and you don't have
a program that understands the header or you want to ignore the header.

.TP
.B -rowskip \fIN
If there is padding at the ends of the rows, you can skip it with this
option.  Note that rowskip need not be an integer.  Amazingly, I once
had an image with 0.376 bytes of padding per row.  This turned out to
be due to a file-transfer problem, but I was still able to read the
image.

Skipping a fractional byte per row means skipping one byte per multiple
rows.
.TP
.B -bt -bottomfirst
By default, 
.B rawtopgm
assumes the pixels in the input go top to bottom, left to right.
If you specify 
.B -bt
or
.BR -bottomfirst ,
.B rawtopgm
assumes the pixels go bottom to top, left to right.  The Molecular Dynamics
and Leica confocal format, for example, use the latter arrangement.

If you don't specify 
.B -bt
when you should or vice versa, the resulting image is upside down, which 
you can correct with 
.B pnmflip .

This option causes 
.B rawtopgm
to read the entire input stream into storage at once, which may take a
lot of storage.  Ordinarly, 
.B rawtopgm
stores only one row at a time.

For backwards compatibility, 
.B rawtopgm
also accepts 
.B -tb 
and 
.B -topbottom
to mean exactly the same thing.  The reasons these are named backwards is that
the original author thought of it as specifying that the wrong results of 
assuming the data is top to bottom should be corrected by flipping the result
top for bottom.  Today, we think of it as simply specifying the format of the
input data so that there are no wrong results.

.SH "SEE ALSO"
.BR pgm (5), 
.BR rawtoppm (1), 
.BR pnmflip (1)

.SH AUTHORS
Copyright (C) 1989 by Jef Poskanzer.
.br
Modified June 1993 by Oliver Trepte, oliver@fysik4.kth.se
.\" Permission to use, copy, modify, and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted, provided
.\" that the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.  This software is provided "as is" without express or
.\" implied warranty.


pgm/rawtopgm.c
/* rawtopgm.c - convert raw grayscale bytes into a portable graymap
**
** Copyright (C) 1989 by Jef Poskanzer.
**
** Permission to use, copy, modify, and distribute this software and its
** documentation for any purpose and without fee is hereby granted, provided
** that the above copyright notice appear in all copies and that both that
** copyright notice and this permission notice appear in supporting
** documentation.  This software is provided "as is" without express or
** implied warranty.
*/

#include <math.h>
#include "pgm.h"


struct cmdline_info {
    /* All the information the user supplied in the command line,
       in a form easy for the program to use.
    */
    char *input_filespec;  /* Filespec of input file */
    unsigned int headerskip;
    float rowskip;
    int bottomfirst;  /* the -bottomfirst/-bt option */
    int autosize;  /* User wants us to figure out the size */
    int width;
    int height;
    int bpp;
      /* bytes per pixel in input format.  1 or 2 */
    int littleendian;
      /* logical: samples in input are least significant byte first */
    int maxval;  /* -maxval option, or -1 if none */
};


static void
parse_command_line(int argc, char ** argv,
                   struct cmdline_info *cmdline_p) {
/*----------------------------------------------------------------------------
   Note that the file spec array we return is stored in the storage that
   was passed to us as the argv array.
-----------------------------------------------------------------------------*/
    optStruct *option_def = malloc(100*sizeof(optStruct));
        /* Instructions to OptParseOptions2 on how to parse our options.
         */
    optStruct2 opt;

    unsigned int option_def_index;

    option_def_index = 0;   /* incremented by OPTENTRY */
    OPTENTRY(0,   "bottomfirst",   OPT_FLAG,   &cmdline_p->bottomfirst,    0);
    OPTENTRY(0,   "bt",            OPT_FLAG,   &cmdline_p->bottomfirst,    0);
    OPTENTRY(0,   "topbottom",     OPT_FLAG,   &cmdline_p->bottomfirst,    0);
    OPTENTRY(0,   "tb",            OPT_FLAG,   &cmdline_p->bottomfirst,    0);
    OPTENTRY(0,   "headerskip",    OPT_UINT,   &cmdline_p->headerskip,     0);
    OPTENTRY(0,   "rowskip",       OPT_FLOAT,  &cmdline_p->rowskip,        0);
    OPTENTRY(0,   "bpp",           OPT_INT,    &cmdline_p->bpp,            0);
    OPTENTRY(0,   "littleendian",  OPT_FLAG,   &cmdline_p->littleendian,   0);
    OPTENTRY(0,   "maxval",        OPT_UINT,   &cmdline_p->maxval,         0);

    /* Set the defaults */
    cmdline_p->bottomfirst = FALSE;
    cmdline_p->headerskip = 0;
    cmdline_p->rowskip = 0.0;
    cmdline_p->bpp = 1;
    cmdline_p->littleendian = 0;
    cmdline_p->maxval = -1;

    opt.opt_table = option_def;
    opt.short_allowed = FALSE;  /* We have no short (old-fashioned) options */
    opt.allowNegNum = FALSE;  /* We may have parms that are negative numbers */

    pm_optParseOptions2(&argc, argv, opt, 0);
        /* Uses and sets argc, argv, and some of *cmdline_p and others. */

    if (argc-1 == 0) {
        cmdline_p->input_filespec = "-";
        cmdline_p->autosize = TRUE;
    } else if (argc-1 == 1) {
        cmdline_p->input_filespec = argv[1];
        cmdline_p->autosize = TRUE;
    } else if (argc-1 == 2) {
        cmdline_p->input_filespec = "-";
        cmdline_p->autosize = FALSE;
        cmdline_p->width = atoi(argv[1]);
        cmdline_p->height = atoi(argv[2]);
    } else if (argc-1 == 3) {
        cmdline_p->input_filespec = argv[3];
        cmdline_p->autosize = FALSE;
        cmdline_p->width = atoi(argv[1]);
        cmdline_p->height = atoi(argv[2]);
    } else
        pm_error("Program takes zero, one, two, or three arguments.  You "
                 "specified %d", argc-1);

    if (cmdline_p->bpp != 1 && cmdline_p->bpp != 2) 
        pm_error("Bytes per pixel (-bpp) must be 1 or 2.  You specified %d.",
                 cmdline_p->bpp);

    if (cmdline_p->maxval == 0)
        pm_error("Maxval (-maxval) may not be zero.");

    if (cmdline_p->maxval > 255 && cmdline_p->bpp == 1)
        pm_error("You have specified one byte per pixel, but a maxval "
                 "too large to fit in one byte: %d", cmdline_p->maxval);
    if (cmdline_p->maxval > 65535)
        pm_error("Maxval must be less than 65536.  You specified %d.",
                 cmdline_p->maxval);

    if (!cmdline_p->autosize && (cmdline_p->width <= 0)) /* #199053 */
        pm_error("Width must be a positive number.");
    if (!cmdline_p->autosize && (cmdline_p->height <= 0)) /* #199053 */
        pm_error("Height must be a positive number.");

    if (cmdline_p->rowskip && cmdline_p->autosize)
        pm_error("If you specify -rowskip, you must also give the image "
                 "dimensions.");
    if (cmdline_p->rowskip && cmdline_p->bottomfirst)
        pm_error("You canot specify both -rowskip and -bottomfirst.  This is "
                 "a limitation of this program.");

}



static void
compute_image_size(const struct cmdline_info cmdline, const long nread,
                   int * const rows_p, int * const cols_p) {

    if (cmdline.autosize) {
        int sqrt_trunc = 
            (int) sqrt((double) (nread-cmdline.headerskip));
        if (sqrt_trunc*sqrt_trunc+cmdline.headerskip != nread) 
            pm_error( "You must specify the dimensions of the image unless "
                      "it is a quadratic image.  This one is not quadratic: "
                      "The number of "
                      "pixels in the input is %ld, which is not a perfect "
                      "square.", nread-cmdline.headerskip);
        *rows_p = *cols_p = sqrt_trunc;
        pm_message( "Image size: %d cols, %d rows", *cols_p, *rows_p);
    } else {
        *rows_p = cmdline.height;
        *cols_p = cmdline.width;
    }
}



static void
skip_header(FILE *ifp, const int headerskip) {
    int i;

    for ( i = 0; i < headerskip; ++i ) {
        /* Read a byte out of the file */
        int val;
        val = getc( ifp );
        if ( val == EOF )
            pm_error("EOF / read error reading Byte %d in the header", i );
    }
}



static gray
read_from_file(FILE *ifp, const int bpp, const int row, const int col,
               const int littleendian) {
/*----------------------------------------------------------------------------
   Return the next sample value from the input file 'ifp', assuming the
   input stream is 'bpp' bytes per pixel (1 or 2).  In the case of two
   bytes, if 'littleendian', assume least significant byte is first.
   Otherwise, assume MSB first.
   
   In error messages, say this is Column 'col', Row 'row'.  Exit program if
   error.
-----------------------------------------------------------------------------*/
    gray retval;

    if (bpp == 1) {
        int val;
        val = getc(ifp);
        if (val == EOF)
            pm_error( "EOF / read error at Row %d Column %d",
                      row, col);
        retval = (gray) val;
    } else {
        short val;
        int rc;
        rc = littleendian ? 
            pm_readlittleshort(ifp, &val) : pm_readbigshort(ifp, &val);
        if (rc != 0)
            pm_error( "EOF / read error at Row %d Column %d",
                      row, col);
        retval = (gray) val;
    }
    return retval;
}



int
main(int argc, char *argv[] ) {

    struct cmdline_info cmdline;
    FILE* ifp;
    gray* grayrow;
    int rows, cols;
    gray maxval;
    char* buf;
    /* pixels_1 and pixels_2 are the array of pixels in the input buffer
       (assuming we are using an input buffer).  pixels_1 is the array
       as if the pixels are one byte each.  pixels_2 is the array as if
       they are two bytes each.
       */
    unsigned char *pixels_1;  
    unsigned short *pixels_2;
    long nread;
    int row;
    float toskip;

    pgm_init( &argc, argv );

    parse_command_line(argc, argv, &cmdline);

    ifp = pm_openr(cmdline.input_filespec);

    if (cmdline.autosize || cmdline.bottomfirst) {
        buf = pm_read_unknown_size( ifp, &nread );
        pixels_1 = (unsigned char *) buf;
        pixels_2 = (unsigned short *) buf;
    } else
        buf = NULL;

    compute_image_size(cmdline, nread, &rows, &cols);

    if (!buf)
        skip_header(ifp, cmdline.headerskip);

    toskip = 0.00001;

    if (cmdline.maxval == -1)
        maxval = (cmdline.bpp == 1 ? (gray) 255 : (gray) 65535);
    else
        maxval = (gray) cmdline.maxval;

    pgm_writepgminit( stdout, cols, rows, maxval, 0 );
    grayrow = pgm_allocrow( cols );

    for ( row = 0; row < rows; ++row) {
        int col;
        unsigned int rowpos; /* index of this row in pixel array */
        if (cmdline.bottomfirst)
            rowpos = (rows-row-1) * cols;
        else
            rowpos = row * cols;

        for ( col = 0; col < cols; ++col )
            if (buf) {
                if (cmdline.bpp == 1)
                    grayrow[col] = pixels_1[rowpos+col];
                else
                    grayrow[col] = pixels_2[rowpos+col];
            } else {
                grayrow[col] = read_from_file(ifp, cmdline.bpp, 
                                              row, col,
                                              cmdline.littleendian);
            }
        for ( toskip += cmdline.rowskip; toskip >= 1.0; toskip -= 1.0 ) {
            /* Note that if we're using a buffer, cmdline.rowskip is zero */
            int val;
            val = getc( ifp );
            if ( val == EOF )
                pm_error( "EOF / read error skipping bytes at the end "
                          "of Row %d.", row);
        }
        pgm_writepgmrow( stdout, grayrow, cols, maxval, 0 );
    }
    
    if (buf)
        free(buf);
    pm_close( ifp );
    pm_close( stdout );

    exit( 0 );
}
